File: familyvault/frontmatter.py
"""Read and write Obsidian-style YAML frontmatter."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

_DELIM = "---"


@dataclass
class Document:
    """A markdown file split into structured frontmatter + body."""

    frontmatter: dict[str, Any]
    body: str

    def render(self) -> str:
        """Serialize to a string suitable for writing to disk."""
        fm = yaml.safe_dump(
            self.frontmatter,
            sort_keys=False,
            allow_unicode=True,
            default_flow_style=False,
        ).rstrip()
        body = self.body.rstrip() + "\n" if self.body.strip() else ""
        return f"{_DELIM}\n{fm}\n{_DELIM}\n\n{body}"


def parse(text: str) -> Document:
    """Parse a markdown string. Missing frontmatter yields an empty dict."""
    lines = text.splitlines()
    if not lines or lines[0].strip() != _DELIM:
        return Document(frontmatter={}, body=text)
    try:
        end = lines.index(_DELIM, 1)
    except ValueError:
        return Document(frontmatter={}, body=text)
    fm_text = "\n".join(lines[1:end])
    body = "\n".join(lines[end + 1 :]).lstrip("\n")
    try:
        fm = yaml.safe_load(fm_text) or {}
    except yaml.YAMLError:
        # Malformed YAML — preserve the original text in the body so nothing is
        # lost, and start with an empty frontmatter. The user can repair the
        # file by hand.
        fm = {}
        body = f"<!-- frontmatter could not be parsed -->\n{fm_text}\n\n{body}"
    if not isinstance(fm, dict):
        fm = {}
    return Document(frontmatter=fm, body=body)


def read(path: Path) -> Document:
    """Read and parse a file from disk."""
    return parse(path.read_text())


def write(path: Path, doc: Document) -> None:
    """Write a Document to disk, creating parents as needed."""
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(doc.render())

File: src/lib/frontmatter.ts
// Read/write Obsidian-style YAML frontmatter. Tolerant of missing or
// malformed frontmatter; preserves key order on render.

import yaml from 'js-yaml';

const DELIM = '---';

export interface Document {
  frontmatter: Record<string, unknown>;
  body: string;
}

export function parse(text: string): Document {
  const lines = text.split(/\r?\n/);
  if (lines.length === 0 || lines[0].trim() !== DELIM) {
    return { frontmatter: {}, body: text };
  }
  const end = lines.indexOf(DELIM, 1);
  if (end === -1) {
    return { frontmatter: {}, body: text };
  }
  const fmText = lines.slice(1, end).join('\n');
  let body = lines.slice(end + 1).join('\n').replace(/^\n+/, '');

  let fm: Record<string, unknown> = {};
  try {
    const loaded = yaml.load(fmText);
    if (loaded && typeof loaded === 'object' && !Array.isArray(loaded)) {
      fm = loaded as Record<string, unknown>;
    }
  } catch {
    body = `<!-- frontmatter could not be parsed -->\n${fmText}\n\n${body}`;
  }
  return { frontmatter: fm, body };
}

export function render(doc: Document): string {
  const fmText = yaml.dump(doc.frontmatter, {
    sortKeys: false,
    lineWidth: -1,
    quotingType: '"',
  }).replace(/\n$/, '');
  const body = doc.body.trim() ? doc.body.replace(/\s+$/, '') + '\n' : '';
  return `${DELIM}\n${fmText}\n${DELIM}\n\n${body}`;
}
